### ISSUE TITLE
Anomaly "File "kernel/inductive.ml", line 1404, characters 12-18: Assertion failed."

### ISSUE THREAD

**Issue by afdw** [coq/coq#17073]
#### Description of the problem

The code
```coq
Inductive vector : nat -> Type :=
  | vector_Nil : vector 0
  | vector_Cons n : vector n -> vector (S n).

Fixpoint bv_and {n} (bv1 bv2 : vector n) : vector n :=
  match bv1 with
  | vector_Nil => fun _ => vector_Nil
  | vector_Cons n1' bv1' => fun bv2 =>
    match (bv2 : vector (S n1')) in (vector n2) return
      (match n2 with 0 => Type | S n2' => (vector n2' -> vector n1') -> vector (S n1') end) with
    | vector_Nil => Type
    | vector_Cons _ bv2' => fun cast => vector_Cons _ (bv_and bv1' (cast bv2'))
    end (fun x => x)
  end bv2.
```
gives the following error:
```
Anomaly "File "kernel/inductive.ml", line 1404, characters 12-18: Assertion failed."
Please report at http://coq.inria.fr/bugs/.
```

#### Coq Version

```
user@notebook:~$ coqtop -v
The Coq Proof Assistant, version 8.16.1
compiled with OCaml 4.14.0
```

**Comment by afdw** [comment]
For information, one way to make that code work is to replace `Type : Type` with `tt : unit`:
```coq
Fixpoint bv_and {n} (bv1 bv2 : vector n) : vector n :=
  match bv1 with
  | vector_Nil => fun _ => vector_Nil
  | vector_Cons n1' bv1' => fun bv2 =>
    match (bv2 : vector (S n1')) in (vector n2) return
      (match n2 with 0 => unit | S n2' => (vector n2' -> vector n1') -> vector (S n1') end) with
    | vector_Nil => tt
    | vector_Cons _ bv2' => fun cast => vector_Cons _ (bv_and bv1' (cast bv2'))
    end (fun x => x)
  end bv2.
  ```

**Comment by ejgallego** [comment]
```
Anomaly
"File "kernel/inductive.ml", line 1404, characters 12-18: Assertion failed."
Please report at http://coq.inria.fr/bugs/.
Raised at Inductive.check_one_fix.check_rec_call_stack in file "kernel/inductive.ml", line 1404, characters 12-40
Called from CArray.fold_left_i.fold in file "clib/cArray.ml", line 233, characters 56-74
Called from Inductive.check_one_fix.check_rec_call_stack in file "kernel/inductive.ml", line 1248, characters 14-192
Called from CArray.fold_left_i.fold in file "clib/cArray.ml", line 233, characters 56-74
Called from Inductive.check_one_fix.check_rec_call_stack in file "kernel/inductive.ml", line 1248, characters 14-192
Called from Inductive.check_one_fix.check_rec_call in file "kernel/inductive.ml", line 1452, characters 19-70
Called from Inductive.check_one_fix in file "kernel/inductive.ml", line 1455, characters 24-50
Called from Inductive.check_fix in file "kernel/inductive.ml", line 1526, characters 10-45
Called from Pretyping.search_guard.(fun) in file "pretyping/pretyping.ml", line 110, characters 14-31
Called from Stdlib__List.iter in file "list.ml", line 110, characters 12-15
Called from Pretyping.search_guard in file "pretyping/pretyping.ml", line 97, characters 7-828
Called from Declare.mutual_make_bodies in file "vernac/declare.ml", line 673, characters 18-77
Called from Declare.declare_mutually_recursive_core in file "vernac/declare.ml", line 685, characters 4-87
Called from ComFixpoint.declare_fixpoint_generic in file "vernac/comFixpoint.ml", line 294, characters 4-126
Called from Vernacentries.vernac_fixpoint in file "vernac/vernacentries.ml", line 1044, characters 13-100
Called from Vernacextend.vtmodifyprogram.(fun) in file "vernac/vernacextend.ml", line 191, characters 29-34
Called from Vernacinterp.interp_typed_vernac in file "vernac/vernacinterp.ml", line 20, characters 20-113
Called from Vernacinterp.interp_control.(fun) in file "vernac/vernacinterp.ml", line 205, characters 24-69
Called from Flags.with_modified_ref in file "lib/flags.ml", line 17, characters 14-17
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Vernacinterp.interp_gen.(fun) in file "vernac/vernacinterp.ml", line 255, characters 18-43
Called from Vernacinterp.interp_gen in file "vernac/vernacinterp.ml", line 253, characters 6-279
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Coq__Interp.coq_interp in file "coq/interp.ml" (inlined), line 35, characters 2-29
Called from Coq__Interp.interp.(fun) in file "coq/interp.ml", line 39, characters 16-34
Called from Coq__Protect.eval in file "coq/protect.ml", line 14, characters 14-17
```

**Comment by SkySkimmer** [comment]
17ade41bb215b47fd49d190cf5873683e3f2bc1e is the first bad commit
cc @herbelin

**Comment by herbelin** [comment]
Fixed in #17116, thanks.
